### DISCUSSION TITLE
Comparison to python-Shiny?

### DISCUSSION THREAD

**Discussion by kshen-noble** [widgetti/solara#711] [Q&A] [answered]
Hi all, 

Really cool work.

I was wondering if there might be users here who could help me understand some of the differences between python-Shiny and Solara, especially as pertains to what can be developed/maintained, app-wise. I couldn't find any other discussions in my searches, but if this exists some where, links would be appreciated, too!

Thanks!

**Comment by maartenbreddels** [comment, marked as answer]
Hi,

good question.
I think there are similarities in how we work with state. Both are reactive, where you don't have to manually add event handers.

In shiny, the state handling is separated from the UI/component/visuals. Some people prefer that, but for instance in their TODO app you see that the state handling code needs to also modify the UI, so this separation doesn't seem to work that well in practice (although vue has done it, but it requires logic in the UI layer: loops, if).

Solara is built on top of ipywidgets, so we render natively on Jupyter platforms. I don't think shiny can do that.

Also, we are strong in typing, our reactive variables (state) is typed, as well as our components, such that if you try to bind a string reactive variable to a component that expects an integer, you'll get a warning from mypy or other type checkers.

Also, we don't use so-called "stringly" typing, where "strings"/"ids" in the UI need to match other strings in the state handling. In solara, we use variables and arguments, if they don't match, it does not pass silently, you will get an exception, but also the linter/type checker will warn before you leave your editor.
 
I hope this helps, let me know if you agree/disagree.

Regards,

Maarten

**Reply by kshen-noble** [reply]
Thank you! That is very helpful, especially the two points regarding typing -- this is a really good case for Solara.

The jupyter integration is also a good point. I've seen demos on DominoAI where they go from solara-jupyter notebooks to laying out dashboards, is this feature specific to Domino's implementation of solara? I couldn't find documentation of this on Solara's web site. I believe [Panel/Holoviz has something like this](https://panel.holoviz.org/how_to/notebook/jupyterlabpreview.html).

Re state handling, I believe shiny.express allows for a more integrated state-handling with UI/component/visuals, instead of the UI/server separation design pattern in `shiny core` and shown in the todo demo that you describe.

**Reply by maartenbreddels** [reply]
We developed that for Domino, but it's not part of the open-source part of solara.
